Source/WTF/wtf/text/StringImpl.cpp

```diff
@@ -1458,6 +1458,8 @@ bool equal(const StringImpl* a, const LChar* b)
 
 bool equal(const StringImpl& a, const StringImpl& b)
 {
+    if (a.hash() != b.hash())
```

Comment:
If you check the implementation of StringImpl's `hash`, it says

```
unsigned hash() const { return hasHash() ? rawHash() : hashSlowCase(); }
```

So, for comparing strings with large sizes, maybe we should check and compare `rawHash()` (the cached hash value) since `hashSlowCase ()` (the hash computation) might be costly.

Reply:
Something like this

```
if (a.hasHash() && b.hasHash()) {
    if (a.rawHash() != b.rawHash())
        return false;
}
```

Reply:
Yes.
1. StringImpl::hash computes the hash and store it, in a non-thread safe way. So we should not always do it in underlying function (DFG may use the other thread's StringImpl only for comparison).
2. hashSlowCase is costly.

Let's do this,
```
unsigned aHash = a.rawHash();
unsigned bHash = b.rawHash();
if (aHash != bHash && aHash && bHash)
    return false;
```

Reply:
Ah, thank you for pointing this out! I still have to A/B test this particular change so it fell under the radar, I'll fix it right away.

Reply:
Resolved in latest push.